### 1 Introduction
### 2 Load and check data
-  load data
- Outlier detection
-  joining train and test set
-  check for null and missing values
### 3 Feature analysis
- 3.1 Numerical values
- 3.2 Categorical values
### 4 Filling missing Values
- 4.1 Age
### 5 Feature engineering
- 5.1 Name/Title
- 5.2 Family Size
- 5.3 Cabin
- 5.4 Ticket
### 6 Modeling
- 6.1 Simple modeling
    - 6.1.1 Cross validate models
    - 6.1.2 Hyperparamater tunning for best models
    - 6.1.3 Plot learning curves
    - 6.1.4 Feature importance of the tree based classifiers
- 6.2 Ensemble modeling
    - 6.2.1 Combining models
- 6.3 Prediction
    - 6.3.1 Predict and Submit results

In [1]:
#import necessary libraries for the project
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
%matplotlib inline

In [2]:
#2.1 load data
train = pd.read_csv("train.csv")
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
test = pd.read_csv('test.csv')
test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [4]:
#2.2 Outlier detection
train.describe(include='all')

PassengerId    Survived      Pclass                         Name  \
count    891.000000  891.000000  891.000000                          891   
unique          NaN         NaN         NaN                          891   
top             NaN         NaN         NaN  Chaffee, Mr. Herbert Fuller   
freq            NaN         NaN         NaN                            1   
mean     446.000000    0.383838    2.308642                          NaN   
std      257.353842    0.486592    0.836071                          NaN   
min        1.000000    0.000000    1.000000                          NaN   
25%      223.500000    0.000000    2.000000                          NaN   
50%      446.000000    0.000000    3.000000                          NaN   
75%      668.500000    1.000000    3.000000                          NaN   
max      891.000000    1.000000    3.000000                          NaN   

         Sex         Age       SibSp       Parch Ticket        Fare    Cabin  \
count    891  714.000000  891.000000  891.000000    891  891.000000      204   
unique     2         NaN         NaN         NaN    681         NaN      147   
top     male         NaN         NaN         NaN   1601         NaN  B96 B98   
freq     577         NaN         NaN         NaN      7         NaN        4   
mean     NaN   29.699118    0.523008    0.381594    NaN   32.204208      NaN   
std      NaN   14.526497    1.102743    0.806057    NaN   49.693429      NaN   
min      NaN    0.420000    0.000000    0.000000    NaN    0.000000      NaN   
25%      NaN   20.125000    0.000000    0.000000    NaN    7.910400      NaN   
50%      NaN   28.000000    0.000000    0.000000    NaN   14.454200      NaN   
75%      NaN   38.000000    1.000000    0.000000    NaN   31.000000      NaN   
max      NaN   80.000000    8.000000    6.000000    NaN  512.329200      NaN   

       Embarked  
count       889  
unique        3  
top           S  
freq        644  
mean        NaN  
std         NaN  
min         NaN  
25%         NaN  
50%         NaN  
75%         NaN  
max         NaN

In [5]:
train['Age'][train['Age'] < 20 ].describe()

count    164.000000
mean      11.979695
std        6.656828
min        0.420000
25%        5.000000
50%       15.000000
75%       18.000000
max       19.000000
Name: Age, dtype: float64

In [6]:
train['Age'][train['Age'] > 65 ].describe()

count     8.000000
mean     71.562500
std       4.048258
min      66.000000
25%      70.000000
50%      70.750000
75%      71.750000
max      80.000000
Name: Age, dtype: float64

In the above 2 cells, I am trying to find outliers. 

The first one print the details of all kids(up to 19 years of age). There are 164 kids in the training dataset the youngest baby is just over months. 

The oldest person in training set is 80 years old and there are 8 people over the age of 65. 

In [7]:
train.loc[train['Age']== 0.420000]

PassengerId  Survived  Pclass                             Name   Sex  \
803          804         1       3  Thomas, Master. Assad Alexander  male   

      Age  SibSp  Parch Ticket    Fare Cabin Embarked  
803  0.42      0      1   2625  8.5167   NaN        C

In [8]:
train.loc[train['Age']== 80]

PassengerId  Survived  Pclass                                  Name  \
630          631         1       1  Barkworth, Mr. Algernon Henry Wilson   

      Sex   Age  SibSp  Parch Ticket  Fare Cabin Embarked  
630  male  80.0      0      0  27042  30.0   A23        S

In [9]:
#print the details of passengers who pay more than average ticket price  of $35
train['Fare'][train['Fare'] > 265 ].describe()
#I am not dropping the rows with highest fare. This may not be an outlier since the passengers Embarked is C, that is Cherbourg in France.
#So that could be the reason for higer fare. 

count      3.0000
mean     512.3292
std        0.0000
min      512.3292
25%      512.3292
50%      512.3292
75%      512.3292
max      512.3292
Name: Fare, dtype: float64

In [10]:
test.describe(include='all')

PassengerId      Pclass                        Name   Sex         Age  \
count    418.000000  418.000000                         418   418  332.000000   
unique          NaN         NaN                         418     2         NaN   
top             NaN         NaN  Widener, Mr. George Dunton  male         NaN   
freq            NaN         NaN                           1   266         NaN   
mean    1100.500000    2.265550                         NaN   NaN   30.272590   
std      120.810458    0.841838                         NaN   NaN   14.181209   
min      892.000000    1.000000                         NaN   NaN    0.170000   
25%      996.250000    1.000000                         NaN   NaN   21.000000   
50%     1100.500000    3.000000                         NaN   NaN   27.000000   
75%     1204.750000    3.000000                         NaN   NaN   39.000000   
max     1309.000000    3.000000                         NaN   NaN   76.000000   

             SibSp       Parch    Ticket        Fare            Cabin Embarked  
count   418.000000  418.000000       418  417.000000               91      418  
unique         NaN         NaN       363         NaN               76        3  
top            NaN         NaN  PC 17608         NaN  B57 B59 B63 B66        S  
freq           NaN         NaN         5         NaN                3      270  
mean      0.447368    0.392344       NaN   35.627188              NaN      NaN  
std       0.896760    0.981429       NaN   55.907576              NaN      NaN  
min       0.000000    0.000000       NaN    0.000000              NaN      NaN  
25%       0.000000    0.000000       NaN    7.895800              NaN      NaN  
50%       0.000000    0.000000       NaN   14.454200              NaN      NaN  
75%       1.000000    0.000000       NaN   31.500000              NaN      NaN  
max       8.000000    9.000000       NaN  512.329200              NaN      NaN

In [11]:
print("Train data: ")
print(train.isna().sum()[(train.isna().sum()!=0)].reset_index())
print("Test data: ")
print(test.isna().sum()[test.isna().sum()!=0].reset_index())

Train data: 
      index    0
0       Age  177
1     Cabin  687
2  Embarked    2
Test data: 
   index    0
0    Age   86
1   Fare    1
2  Cabin  327


Training and testing dataset has missing values in 3 columns each. Both these datasets has missing values in Age and Cabin columns. 

In my opinion we can safely remove cabin data because there are so many missing items(more 77% in training and 78% testing data missing). 
In case of Age, nearly 20% data missing for training and testing set. But Age can be an important feature so I don't want to drop that from our dataset. I would like to impute value later. 

Other 2 missing values are Embarked in training and Fare in testing data. less than 1% data missing, so we can keep that data and impute values later. In my opinion both Embarked and Fare column value is insignificant. Let me analyze further to conclude that. 

In [12]:
train = train.drop(['Cabin'], axis=1)
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

In [13]:
test = test.drop(['Cabin'], axis=1)
test.shape

(418, 10)

Encode values for 'Age', 'Embarked' and 'Fare' columns in both train and testing sets. 

In [14]:
#Encode values for Age - Age is a number so we can fill the missing columns with the mean age value.
train['Age'] = train['Age'].fillna(round( train['Age'].mean()))
mode_em = train['Embarked'].mode()
train['Embarked'] = train['Embarked'].fillna(train['Embarked'].mode()[0])
print(train.isna().sum())

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64


In [17]:
#We can encode the values of testing set as well 
test['Age'] = test['Age'].fillna(round(test['Age'].mean()))
test['Fare'] = test['Fare'].fillna(round(test['Fare'].mean()))
print(test.isna().sum())

PassengerId    0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64


In [27]:
#Now lets analyze data - how people survived on different Pclass
train[['Pclass', 'Survived', 'Fare']].groupby(['Pclass'], as_index=False).mean()

Pclass  Survived       Fare
0       1  0.629630  84.154687
1       2  0.472826  20.662183
2       3  0.242363  13.675550

In [36]:
train[['Sex', 'Survived']].groupby(['Sex'], as_index=False).count()

Sex  Survived
0  female       314
1    male       577

In [38]:
train[['Age', 'Survived']].groupby(['Survived'], as_index=False).mean()

Survived        Age
0         0  30.483607
1         1  28.595526

In [39]:
train[['Embarked', 'Survived']].groupby(['Embarked'], as_index=False).count()

Embarked  Survived
0        C       168
1        Q        77
2        S       646